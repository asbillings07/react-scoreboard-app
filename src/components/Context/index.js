import React, { Component } from "react";

// Step 1: creating context and assigning to a variable
// steps up provider and consumer.
const ScoreboardContext = React.createContext();
// pulling out Provider and Consumer from the context and exporting them to use in other places.
/*******
 Higher Order Component - A techinque in React for resuing component logic. A HOC takes an exisiting component and returns a new component. Acts as a wrapper that takes an existing componet and returns a new component. 

 Whereas a component transforms props into UI, a higher-order component transforms a component into another component.
 ************/
export class Provider extends Component {
  state = {
    players: [
      {
        name: "Aaron",
        score: 0,
        id: 1
      },
      {
        name: "Amanda",
        score: 0,
        id: 2
      },
      {
        name: "Esther",
        score: 0,
        id: 3
      },
      {
        name: "Eden",
        score: 0,
        id: 4
      }
    ]
  };
  //Player id counter
  prevPlayerId = 4;

  handleScoreChange = (index, delta) => {
    this.setState(prevState => ({
      score: (prevState.players[index].score += delta)
    }));
  };

  handleRemovePlayer = id => {
    this.setState(prevState => {
      return {
        players: prevState.players.filter(p => p.id !== id)
      };
    });
  };

  handleAddPlayer = name => {
    this.setState(prevState => {
      return {
        players: [
          ...prevState.players,
          {
            name,
            score: 0,
            id: (this.prevPlayerId += 1)
          }
        ]
      };
    });
  };

  getHighScore = () => {
    const scores = this.state.players.map(p => p.score);
    const highScore = Math.max(...scores);

    if (highScore) {
      return highScore;
    } else {
      return null;
    }
  };

  render() {
    return (
      <ScoreboardContext.Provider
        value={{
          players: this.state.players,
          actions: {
            changeScore: this.handleScoreChange,
            removePlayer: this.handleRemovePlayer,
            addPlayer: this.handleAddPlayer,
            highScore: this.getHighScore()
          }
        }}
      >
        {this.props.children}
        {}
      </ScoreboardContext.Provider>
    );
  }
}

export const Consumer = ScoreboardContext.Consumer;

/* this.props.children - A special React prop that lets you pass components as data to other componenets.This allows the App component along with all it's children to be passed into the Provider and rendered. */

/**
 * 
 * 
 * Context API:

Consists of 3 parts:

1. createContext() method
    1. Sets up a context and returns an object with a Provider and a Consumer
2. Provider
    1. Used as high as possible in the tree. It allows a Consumer to subscribe to context changes
    2. The provider component requires a Value Prop, actions or event handlers. 
    3. Any desendant of the provider componenet will have access to the data given to the value Prop.

3. Consumer 
    1. Access the Provider to get any data needed. The Consumer helps avoid prop drilling.
    2. Render Prop - A technique for sharing code between React components using a prop whose value is a function. 

Communication between the Provider and Consumer makes context work.

 * 
 * 
 */
