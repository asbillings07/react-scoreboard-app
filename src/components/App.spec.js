import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { Provider } from './Context/index';
import App from './App';

afterEach(cleanup);

test('<App />', () => {
  const { container } = render(
    <Provider>
      <App />
    </Provider>
  );
  expect(container.firstChild).toMatchSnapshot();
});
