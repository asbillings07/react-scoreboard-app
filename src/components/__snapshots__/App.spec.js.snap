// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> 1`] = `
<div
  class="scoreboard"
>
  <header>
    <table
      class="stats"
    >
      <tbody>
        <tr>
          <td>
            Players:
          </td>
          <td>
            4
          </td>
        </tr>
        <tr>
          <td>
            Total Points:
          </td>
          <td>
            0
          </td>
        </tr>
      </tbody>
    </table>
    <h1
      aria-label="the title of the app is"
    >
      Scoreboard
    </h1>
    <div
      class="stopwatch"
    >
      <h2>
        Stopwatch
      </h2>
      <span
        class="stopwatch-time"
      >
        0
      </span>
      <button>
        Start
      </button>
      <button>
        Reset
      </button>
    </div>
  </header>
  <div
    class="player"
  >
    <span
      class="player-name"
    >
      <button
        class="remove-player"
      >
        ✖
      </button>
      <svg
        viewBox="0 0 44 35"
      >
        <path
          d="M26.7616 10.6207L21.8192 0L16.9973 10.5603C15.3699 14.1207 10.9096 15.2672 7.77534 12.9741L0 7.24138L6.56986 28.8448H37.0685L43.5781 7.72414L35.7425 13.0948C32.6685 15.2672 28.3288 14.0603 26.7616 10.6207Z"
          transform="translate(0 0.301727)"
        />
        <rect
          width="30.4986"
          height="3.07759"
          transform="translate(6.56987 31.5603)"
        />
      </svg>
      Aaron
    </span>
    <div
      class="counter"
    >
      <button
        class="counter-action decrement"
      >
        -
      </button>
      <span
        class="counter-score"
      >
        0
      </span>
      <button
        class="counter-action increment"
      >
        +
      </button>
    </div>
  </div>
  <div
    class="player"
  >
    <span
      class="player-name"
    >
      <button
        class="remove-player"
      >
        ✖
      </button>
      <svg
        viewBox="0 0 44 35"
      >
        <path
          d="M26.7616 10.6207L21.8192 0L16.9973 10.5603C15.3699 14.1207 10.9096 15.2672 7.77534 12.9741L0 7.24138L6.56986 28.8448H37.0685L43.5781 7.72414L35.7425 13.0948C32.6685 15.2672 28.3288 14.0603 26.7616 10.6207Z"
          transform="translate(0 0.301727)"
        />
        <rect
          width="30.4986"
          height="3.07759"
          transform="translate(6.56987 31.5603)"
        />
      </svg>
      Amanda
    </span>
    <div
      class="counter"
    >
      <button
        class="counter-action decrement"
      >
        -
      </button>
      <span
        class="counter-score"
      >
        0
      </span>
      <button
        class="counter-action increment"
      >
        +
      </button>
    </div>
  </div>
  <div
    class="player"
  >
    <span
      class="player-name"
    >
      <button
        class="remove-player"
      >
        ✖
      </button>
      <svg
        viewBox="0 0 44 35"
      >
        <path
          d="M26.7616 10.6207L21.8192 0L16.9973 10.5603C15.3699 14.1207 10.9096 15.2672 7.77534 12.9741L0 7.24138L6.56986 28.8448H37.0685L43.5781 7.72414L35.7425 13.0948C32.6685 15.2672 28.3288 14.0603 26.7616 10.6207Z"
          transform="translate(0 0.301727)"
        />
        <rect
          width="30.4986"
          height="3.07759"
          transform="translate(6.56987 31.5603)"
        />
      </svg>
      Esther
    </span>
    <div
      class="counter"
    >
      <button
        class="counter-action decrement"
      >
        -
      </button>
      <span
        class="counter-score"
      >
        0
      </span>
      <button
        class="counter-action increment"
      >
        +
      </button>
    </div>
  </div>
  <div
    class="player"
  >
    <span
      class="player-name"
    >
      <button
        class="remove-player"
      >
        ✖
      </button>
      <svg
        viewBox="0 0 44 35"
      >
        <path
          d="M26.7616 10.6207L21.8192 0L16.9973 10.5603C15.3699 14.1207 10.9096 15.2672 7.77534 12.9741L0 7.24138L6.56986 28.8448H37.0685L43.5781 7.72414L35.7425 13.0948C32.6685 15.2672 28.3288 14.0603 26.7616 10.6207Z"
          transform="translate(0 0.301727)"
        />
        <rect
          width="30.4986"
          height="3.07759"
          transform="translate(6.56987 31.5603)"
        />
      </svg>
      Eden
    </span>
    <div
      class="counter"
    >
      <button
        class="counter-action decrement"
      >
        -
      </button>
      <span
        class="counter-score"
      >
        0
      </span>
      <button
        class="counter-action increment"
      >
        +
      </button>
    </div>
  </div>
  <form>
    <input
      type="text"
      placeholder="Enter a Player's Name"
    />
    <input
      type="submit"
      value="Add Player"
    />
  </form>
</div>
`;
